Allow custom extension for brotli files

Frank3K opened the pull request:
As discussed [here](https://kevinlocke.name/bits/2016/01/20/serving-pre-compressed-files-with-apache-multiviews/) there is no official extension for brotli files yet. Furthermore files with the `.br` extension introduce issues in Apache since `.br` is the [ISO 639 language code](https://www.loc.gov/standards/iso639-2/php/code_list.php) for Breton. 

Therefore I have added support for custom extensions.

This PR includes:
- Necessary code changes
- A test for the added functionality
- Updated README file

I have used more or less the same approach as the [gulp-gzip](https://github.com/jstuckey/gulp-gzip) plugin.

Frank3K committed:
Allow custom extension for brotli files

nstepien commented on README.md:
```diff
@@ -29,7 +39,29 @@ gulp.task('example', function() {
+  return gulp.src('path/to/input')
+    .pipe(brotli.compress({
+      extension: 'brotli',
+      brotliParams: {
```
You're moving `brotliParams` inside another object, that's a breaking change so I'd rather the `extension` property was added to the `brotliParams` object instead.

Frank3K replied:
It is indeed a breaking change. This separation was done on purpose to distinguish between parameters for the plugin and parameters for iltorb

If `extension` is added to the `brotliParams` object it would also be passed to iltorb. This can give unexpected results in the future. Of course we can delete the property before passing the object, but that would exclude the use of a property with the same name in any future iltorb release.

I looked at `gulp-gzip` on how it's handled there. That plugin already has a number of miscellaneous properties for the plugin itself.

My first idea was the same as yours and since the odds are small that a name conflict arises, I could change this PR if you want. What do you think: should I delete the extension property before passing it to iltorb?

nstepien replied:
iltorb is only doing bindings for brotli, and I doubt brotli will ever add an `extension` parameter so it should be fine.

Deleting the property is not an option unless we clone the object, as it could have unexpected side-effects, consider this:

``` js
var brotliParams = {...};

gulp.task('example', function() {
  return gulp.src('path/to/input')
    .pipe(brotli.compress(brotliParams))
    .pipe(gulp.dest('path/to/output'));
});
```

The second time the `example` task is executed, when doing continuous builds for example, the `extension` property will be gone.

Leaving the `extension` property on the object is likely fine, but if we want to be truly future-proof we could do this:

``` js
var brotliParams = {
  // brotli parameters
  quality: 11,
  // gulp parameters
  gulpParams: {
    extension: 'brotli'
  }
};
```

But it would be extra verbose for the decompression part.

I'd go with a flat object for now, and if we really need to we can do a breaking change and adopt proper splitting like `gulp-gzip` in a future version.

gulp itself does just that and gets away with it: "[gulp supports all options supported by node-glob and glob-stream except ignore and adds the following options](https://github.com/gulpjs/gulp/blob/master/docs/API.md#options)"

nstepien commented on index.js:
```diff
@@ -5,9 +5,13 @@ exports.decompress = decompress;
 
 function compress(params) {
-  params = params || {};
+  params = extend(true, {
```
I don't think this is complex enough to warrant an extra dependency.

We can just use this here and in `decompress` as well.

``` js
function getExtension(params) {
  return params && params.extension || 'br';
}

// ...

var extension = getExtension(params);
```

Frank3K replied:
True. I added `extend` such that extra parameters can be added easily in the future. I will change this.

MayhemYDG commented:
While it's true there is no official extension yet, `.br. already has momentum:
- https://github.com/madler/brotli/blob/master/br-format-v3.txt<br>
  There's this proposal for the brotli framing format using `.br`.
- https://github.com/google/ngx_brotli#brotli_static<br>
  Google's brotli module for nginx checks for the `.br` file extension.

MayhemYDG commented:
This also needs to document that `decompress` can accept an optional options object.

```markdown
### compress([options])
#### Default compression
#### Custom compression
### decompress([options])
#### Default decompression
#### Custom decompression
```

Frank3K commented:
Thanks for your review. I have updated the PR based on your feedback and squashed my commits. 

Since the `extension` now is a subproperty of the options object the `README` has been reworded. I'm not entirely satisfied with how it is now though. Feel free to change this or suggest changes.

nstepien commented on index.js:
```diff
@@ -36,14 +42,16 @@ function compress(params) {
-function decompress() {
+function decompress(params) {
+  var extension = getExtension(params);
+  var reExtension = new RegExp('.' + extension + '$', 'i');
```
I missed this the first time but this needs to be `'\\.'` instead of `'.'`.

MayhemYDG commented:
Thanks! I'll do the last fixes myself.
